File: 0013-roman-to-integer/0013-roman-to-integer.cpp
class Solution
{
public:
    int romanToInt(string s)
    {
        int i, ans = 0;
        for (i = 0; i < s.length(); i++)
        {
            switch (s[i])
            {
            case 'I':
                if (s[i + 1] == 'V')
                {
                    ans += 4;
                    i++;
                    break;
                }
                if (s[i + 1] == 'X')
                {
                    ans += 9;
                    i++;
                    break;
                }
                ans += 1;
                break;
            case 'V':
                ans += 5;
                break;
            case 'X':
                if (s[i + 1] == 'L')
                {
                    ans += 40;
                    i++;
                    break;
                }
                if (s[i + 1] == 'C')
                {
                    ans += 90;
                    i++;
                    break;
                }
                ans += 10;
                break;
            case 'L':
                ans += 50;
                break;
            case 'C':
                if (s[i + 1] == 'D')
                {
                    ans += 400;
                    i++;
                    break;
                }
                if (s[i + 1] == 'M')
                {
                    ans += 900;
                    i++;
                    break;
                }
                ans += 100;
                break;
            case 'D':
                ans += 500;
                break;
            case 'M':
                ans += 1000;
                break;
            default:
                break;
            }
        }

        return ans;
    }
};

File: 0013-roman-to-integer/0013-roman-to-integer.py
# time complexity: O(n)
# space complexity: O(1)
class Solution:
    def romanToInt(self, s: str) -> int:
        total = 0
        i = 0
        values = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000,
        }

        for i in range(len(s)):
            if i + 1 < len(s) and values[s[i]] < values[s[i + 1]]:
                total -= values[s[i]]
            else:
                total += values[s[i]]
        return total


s = "III"
print(Solution().romanToInt(s))
s = "LVIII"
print(Solution().romanToInt(s))
s = "MCMXCIV"
print(Solution().romanToInt(s))
